File: examples/langgraph_agent.ts
/**
 * LangGraph Agent — traced by OrchestraAI (TypeScript).
 *
 * Uses the LangChain callback handler injected into LangGraph's config.
 *
 * Prerequisites:
 *   npm link ./sdks/typescript
 *   npm install @langchain/openai @langchain/langgraph
 *
 * Run:
 *   export ORCHESTRA_API_KEY=oai_...
 *   export OPENAI_API_KEY=sk-...
 *   npx tsx examples/langgraph_agent.ts
 */

import { OrchestraAI } from '../sdks/typescript/src';
import { createLangChainHandler } from '../sdks/typescript/src/integrations/langchain';
import { ChatOpenAI } from '@langchain/openai';
import { createReactAgent } from '@langchain/langgraph/prebuilt';
import { tool } from '@langchain/core/tools';
import { z } from 'zod';

const oa = new OrchestraAI({
  apiKey: process.env.ORCHESTRA_API_KEY!,
  baseUrl: process.env.ORCHESTRA_BASE_URL || 'http://localhost:3001',
});

const handler = createLangChainHandler({
  client: oa,
  agentName: 'langgraph-ts-agent',
});

// Define tools
const searchDocs = tool(
  async ({ query }: { query: string }) => {
    return `Found 3 documents about '${query}': [doc1, doc2, doc3]`;
  },
  {
    name: 'search_docs',
    description: 'Search the knowledge base',
    schema: z.object({ query: z.string() }),
  }
);

const getStockPrice = tool(
  async ({ symbol }: { symbol: string }) => {
    const prices: Record<string, string> = { AAPL: '$178.50', GOOGL: '$141.20' };
    return prices[symbol.toUpperCase()] || `No data for ${symbol}`;
  },
  {
    name: 'get_stock_price',
    description: 'Get stock price for a ticker',
    schema: z.object({ symbol: z.string() }),
  }
);

async function main() {
  console.log('=== LangGraph Agent with OrchestraAI Tracing (TypeScript) ===\n');

  const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });
  const agent = createReactAgent({ llm, tools: [searchDocs, getStockPrice] });

  // Pass callbacks via config — OrchestraAI captures all spans
  const result = await agent.invoke(
    { messages: [{ role: 'user', content: 'Search for AI agents info, then check AAPL stock price' }] },
    { callbacks: [handler] },
  );

  for (const msg of result.messages) {
    const content = msg.content;
    if (content) console.log(`  [${msg._getType()}] ${String(content).slice(0, 200)}`);
  }

  console.log('\nCheck traces at http://localhost:3000/dashboard/traces');
}

main().catch(console.error);

File: examples/langgraph_agent.py
"""
LangGraph ReAct Agent — auto-instrumented by OrchestraAI.

Uses `auto_instrument()` to patch LangGraph — no manual callback wiring needed.
All graph invocations (invoke, stream, ainvoke, astream) are traced automatically.

Prerequisites:
    pip install -e sdks/python
    pip install langchain-openai langgraph

Run:
    export ORCHESTRA_API_KEY=oai_...
    export OPENAI_API_KEY=sk-...
    python examples/langgraph_agent.py
"""

import os
from orchestra_ai import OrchestraAI
from orchestra_ai.integrations import langgraph_tracer

from langchain_openai import ChatOpenAI
from langchain_core.tools import tool
from langgraph.prebuilt import create_react_agent

# ── Setup ──────────────────────────────────────────────────────
oa = OrchestraAI(
    api_key=os.environ["ORCHESTRA_API_KEY"],
    base_url=os.getenv("ORCHESTRA_BASE_URL", "http://localhost:3001"),
)

# One line — patches all LangGraph graph methods globally
langgraph_tracer.auto_instrument(oa)


# ── Tools ──────────────────────────────────────────────────────
@tool
def search_docs(query: str) -> str:
    """Search the knowledge base for relevant documents."""
    return f"Found 3 documents about '{query}': [doc1: Overview...] [doc2: Details...] [doc3: Examples...]"


@tool
def get_stock_price(symbol: str) -> str:
    """Get the current stock price for a ticker symbol."""
    prices = {"AAPL": "$178.50", "GOOGL": "$141.20", "MSFT": "$415.30"}
    return prices.get(symbol.upper(), f"No data for {symbol}")


# ── Build Graph ────────────────────────────────────────────────
llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)
agent = create_react_agent(llm, tools=[search_docs, get_stock_price])

# ── Run (automatically traced!) ────────────────────────────────
print("=== LangGraph Agent with OrchestraAI Auto-Instrumentation ===\n")

# invoke() is auto-traced
result = agent.invoke({
    "messages": [{"role": "user", "content": "Search for information about AI agents, then check the stock price of AAPL"}]
})

# Print the final message
for msg in result["messages"]:
    role = getattr(msg, "type", "unknown")
    content = getattr(msg, "content", str(msg))
    if content:
        print(f"  [{role}] {content[:200]}")

# stream() is also auto-traced
print("\n--- Streaming mode ---\n")
for chunk in agent.stream({
    "messages": [{"role": "user", "content": "What's the stock price of GOOGL?"}]
}):
    for key, value in chunk.items():
        print(f"  [{key}] {str(value)[:100]}")

print("\nCheck traces at http://localhost:3000/dashboard/traces")

# Clean up
langgraph_tracer.remove_instrumentation()
oa.close()
